qiita_client/plugin.py

```diff
@@ -0,0 +1,330 @@
+        self.required_parameters = required_parameters
+        self.optional_parameters = optional_parameters
+        self.default_parameter_sets = default_parameter_sets
+        self.outputs = outputs
+
+    def __call__(self, qclient, server_url, job_id, output_dir):
+        return self.function(qclient, server_url, job_id, output_dir)
+
+
+class QiitaArtifactType(object):
+    """A Qiita artifact type
+
+        Parameters
+        ----------
+        name : str
+            The artifact type name
+        description : str
+            The artifact type description
+        can_be_submitted_to_ebi : bool
+            Whether the artifact type can be submitted to EBI or not
+        can_be_submitted_to_vamps : bool
+            Whether the artifact type can be submitted to VAMPS or not
+        filepath_types : list of (str, bool)
+            The list filepath types that the new artifact type supports, and
+            if they're required or not in an artifact instance of this type"""
+    def __init__(self, name, description, can_be_submitted_to_ebi,
+                 can_be_submitted_to_vamps, filepath_types):
+        self.name = name
+        self.description = description
+        self.ebi = can_be_submitted_to_ebi
+        self.vamps = can_be_submitted_to_vamps
+        self.fp_types = filepath_types
+
+
+class BaseQiitaPlugin(object):
+    def __init__(self, name, version, description, publications=None):
+        self.name = name
+        self.version = version
+        self.description = description
+        self.publications = dumps(publications) if publications else ""
+
+        # Will hold the different commands
+        self.task_dict = {}
+
+        # The configuration file
+        conf_dir = environ.get(
+            'QIITA_PLUGINS_DIR', join(expanduser('~'), '.qiita_plugins'))
+        self.conf_fp = join(conf_dir, "%s_%s.conf" % (self.name, self.version))
+
+    def generate_config(self, env_script, start_script, server_cert=None):
+        """Generates the plugin configuration file"""
+        sr = SystemRandom()
+        chars = ascii_letters + digits
+        client_id = ''.join(sr.choice(chars) for i in range(50))
+        client_secret = ''.join(sr.choice(chars) for i in range(255))
+
+        server_cert = server_cert if server_cert else ""
+
+        with open(self.conf_fp, 'w') as f:
+            f.write(CONF_TEMPLATE % (self.name, self.version, self.description,
+                                     env_script, start_script,
+                                     self._plugin_type, self.publications,
+                                     server_cert, client_id, client_secret))
+
+    def _register_command(self, command):
+        """Registers a command in the plugin
+
+        Parameters
+        ----------
+        command: QiitaCommand
+            The command to be added to the plugin
+        """
+        self.task_dict[command.name] = command
+
+    def _install(self, qclient):
```

Comment:
this is semantic, but i think of install as actually putting a program, library, or documentation somewhere in a logical place. in this case, this I think more mirrors a registration, where a call to `._register` would establish this plugins existence with qiita?